android/libraries/rib-badoo/src/main/java/com/badoo/common/rib/RouterConfigurationManager.kt

```diff
@@ -0,0 +1,93 @@
+import com.badoo.common.rib.RouterConfigurationManager.Wish.Pop
+import com.badoo.common.rib.RouterConfigurationManager.Wish.Push
+import com.badoo.common.rib.RouterConfigurationManager.Wish.NewRoot
+import com.badoo.common.rib.RouterConfigurationManager.Wish.Replace
+import com.badoo.mvicore.element.Actor
+import com.badoo.mvicore.element.Reducer
+import com.badoo.mvicore.element.TimeCapsule
+import com.badoo.mvicore.feature.ActorReducerFeature
+import io.reactivex.Observable
+import io.reactivex.Observable.empty
+import io.reactivex.Observable.just
+import kotlinx.android.parcel.Parcelize
+
+class RouterConfigurationManager<C : Parcelable>(
+    initialConfiguration: C,
+    timeCapsule: TimeCapsule<State<C>>,
+    tag: String
+): ActorReducerFeature<Wish<C>, Effect<C>, State<C>, Nothing>(
+    initialState = timeCapsule[tag] ?: State(listOf(initialConfiguration)),
+    actor = ActorImpl<C>(),
+    reducer = ReducerImpl<C>()
+) {
+    init {
+        timeCapsule.register(tag) { state }
+    }
+
+    @Parcelize
+    data class State<C : Parcelable>(
+        val backStack: List<C>
+    ) : Parcelable {
+        val current: C
+            get() = backStack.last()
+
+        val canPop: Boolean
+            get() = backStack.size > 1
+    }
+
+    sealed class Wish<C : Any> {
+        data class Replace<C : Any>(val configuration: C) : Wish<C>()
+        data class Push<C : Any>(val configuration: C) : Wish<C>()
+        data class NewRoot<C : Any>(val configuration: C) : Wish<C>()
+        class Pop<C : Any> : Wish<C>()
+    }
+
+    sealed class Effect<C : Any> {
+        data class Replace<C : Any>(val configuration: C) : Effect<C>()
+        data class Push<C : Any>(val configuration: C) : Effect<C>()
+        data class NewRoot<C : Any>(val configuration: C) : Effect<C>()
+        class Pop<C : Any> : Effect<C>()
+    }
+
+    class ActorImpl<C : Parcelable> : Actor<State<C>, Wish<C>, Effect<C>> {
+        override fun invoke(state: State<C>, wish: Wish<C>): Observable<out Effect<C>> =
+            when (wish) {
+                is Replace -> when {
+                    wish.configuration != state.backStack.last() -> just(Effect.Replace(wish.configuration))
+                    else -> empty()
+                }
+                is Push -> when {
+                    wish.configuration != state.backStack.last() -> just(Effect.Push(wish.configuration))
+                    else -> empty()
+                }
+                is NewRoot-> just(Effect.NewRoot(wish.configuration))
+                is Pop -> when {
+                    state.canPop -> just(Effect.Pop())
+                    else -> empty()
+                }
+            }
+    }
+
+    class ReducerImpl<C : Parcelable> : Reducer<State<C>, Effect<C>> {
+        override fun invoke(state: State<C>, effect: Effect<C>): State<C> = when (effect) {
+            is Effect.Replace -> state.copy(
+                backStack = state.backStack.dropLast(1) + effect.configuration
```

Comment:
What is a backStack is empty?

Reply:
It's guaranteed by Actor to not be empty at this point.